user_tools/src/spark_rapids_pytools/rapids/bootstrap.py

```diff
@@ -0,0 +1,152 @@
+        gpu_mem = worker_info.gpu_info.gpu_mem
+        num_cpus = worker_info.sys_info.num_cpus
+        cpu_mem = worker_info.sys_info.cpu_mem
+
+        constants = self.ctxt.get_value('local', 'clusterConfigs', 'constants')
+        executors_per_node = num_gpus
+        num_executor_cores = max(1, num_cpus // executors_per_node)
+        gpu_concurrent_tasks = min(constants.get('maxGpuConcurrent'), gpu_mem // constants.get('gpuMemPerTaskMB'))
+        # account for system overhead
+        usable_worker_mem = max(0, cpu_mem - constants.get('systemReserveMB'))
+        executor_container_mem = usable_worker_mem // executors_per_node
+        # reserve 10% of heap as memory overhead
+        max_executor_heap = max(0, int(executor_container_mem * (1 - constants.get('heapOverheadFraction'))))
+        # give up to 2GB of heap to each executor core
+        executor_heap = min(max_executor_heap, constants.get('heapPerCoreMB') * num_executor_cores)
+        executor_mem_overhead = int(executor_heap * constants.get('heapOverheadFraction'))
+        # pinned memory uses any unused space up to 4GB
+        pinned_mem = min(constants.get('maxPinnedMemoryMB'),
+                         executor_container_mem - executor_heap - executor_mem_overhead)
+        executor_mem_overhead += pinned_mem
+        res = {
+            'spark.executor.cores': num_executor_cores,
+            'spark.executor.memory': f'{executor_heap}m',
+            'spark.executor.memoryOverhead': f'{executor_mem_overhead}m',
+            'spark.rapids.sql.concurrentGpuTasks': gpu_concurrent_tasks,
+            'spark.rapids.memory.pinnedPool.size': f'{pinned_mem}m',
+            'spark.sql.files.maxPartitionBytes': f'{constants.get("maxSqlFilesPartitionsMB")}m',
+            'spark.task.resource.gpu.amount': 1 / num_executor_cores
+        }
+        return res
+
+    def _run_rapids_tool(self):
+        """
+        Run the bootstrap on the driver node
+        :return:
+        """
+        self.logger.info('Executing Bootstrap commands on remote cluster to calculate default configurations.')
+        exec_cluster: ClusterBase = self.get_exec_cluster()
+        worker_hw_info = exec_cluster.get_worker_hw_info()
+        self.logger.debug('Worker hardware INFO %s', worker_hw_info)
+        try:
+            spark_settings = self.__calculate_spark_settings(worker_info=worker_hw_info)
+            self.ctxt.set_ctxt('bootstrap_results', spark_settings)
+            self.logger.debug('%s Tool finished calculating recommended Apache Spark configurations for cluster %s: %s',
+                              self.pretty_name(),
+                              self.cluster,
+                              str(spark_settings))
+        except Exception as e:
+            self.logger.error('Error while calculating spark configurations')
+            raise e
+
+    def _apply_changes_to_remote_cluster(self):
+        ssh_cmd = "\"sudo bash -c 'cat >> /etc/spark/conf/spark-defaults.conf'\""
```

Comment:
I haven't figured out the whole picture of this PR, but will this `/etc/spark/conf/spark-defaults.con` path also take effect on EMR? (I tried to search this override one for this method, but didn't find it in the PR, so that means it should apply for both dataproc and emr?)

Reply:
yes, Dataproc code is not affected by the changes in this PR.
it is kept in the old package `spark_rapids_dataproc_tools`
This PR creates new package `spark_rapids_pytools` which is all about EMR.